JP Onboarding: Connect Site Title Step to Redux

ockham opened the pull request:
Allow the user to actually set site title and description.

Proposal/RFC. Some notes:
* I've added the site slug explicitly to the route. Even tho we cannot rely on the current `siteSelection`/ `getSelectedSiteId` technique that we employ for _connected_ JP sites in case of unconnected ones, it makes sense to explicitly include the site slug in the route, since a user could have more than one JP sites, and we need to know for which one they're going thru the Onboarding flow.
* This obviously can only work for connected sites as of yet. However, for those, we can use existing selectors and actions from Redux state (`state/site-settings`).
* This means changes are applied immediately upon clicking 'Next Step' (as opposed to e.g. carrying them along and only applying them after the user has finished the JPO flow).
* This has the advantage that the user can navigate back (after navigation is fixed, see testing notes), find the input fields prepopulated with the readily changed values, and can modify them again (e.g. to fix a typo). That's good UX I think.
* This also means we'll have to add logic for the unconnected case. Hopefully we can come up with a generic solution that we can scale to other settings, too. One option would be to extend actions (and selectors/reducers?) to accept not only a site ID (which is a concept inherent to connected sites) but also a site slug, and use whatever auth technique to communicate with the site directly, i.e. bypassing WP.com.

To test:

* Navigate to http://calypso.localhost:3000/jetpack/onboarding/site-title/:site (Note you need to specify a site now!)
* Wait for the form to load. Verify the input fields are prepopulated with the current title and description of the JP site.
* Change title and/or description, click 'Next Step'. Check the site (frontend or wp-admin) and verify that changes have been applied.
* Careful, the 'Back' and 'Skip for now' navigation links don't work right now because of the route change that now includes the site slug!

AnnaMag commented:
Thanks @ockham! 
Quick questions that came to my mind (mostly for my clarity:) ) :
- when the mechanism will be working for non-connected sites, will the "connected-sites" logic be needed for this iteration at all? 
- we 'should' be displaying any site only if `ifSiteNotOnboarded` is `true`, right? Should we start thinking about having a controller on our end and including this in the logic? 

I will have an in-depth read through a bit later.

ockham commented:
Okay, since the data connecting part is still a bit controversial, I'll file a PR to solely add the site slug to the routes for now. Let's continue the data connecting conversation -- either here or elsewhere 😄 I'll start by addressing feedback here :slightly_smiling_face: 

> However, I'm really hesitant to continue with the typical site-object-centric approach - a main requirement is to use the JPO flow for non-connected sites when not logged in, and I don't see how this requirement can be matched with this approach.

> I might be missing something here, though - let me know if that's the case.

One thing I'll point out is that if we agree that we want per-site routes, then we will also want to store data in our Redux tree on a per-site basis. So IMO it boils down to at which layer we seek unification of interfaces, or whether we want to have that at all.

(I agree that the site objects we store for WP.com and connected JP sites tend to be bloated and possibly overkill, but I'd consider that more of an implementation detail.)

Let's say we're storing data for unconnected JP sites in a Redux subtree (let's say we call it `onboarding`) that's entirely distinct from the WP.com/connected JP one, but still on a site slug basis. Wouldn't the duplication of e.g. where a site's title and description are stored be confusing to developers? More so, if this distinction is not just in the tree (i.e. reducer) shape, but also exposed thru the presence of distinct actions for each case?

From an API POV, I, as the developer, would ideally like to use the same action to set a site title, regardless of whether the site is a WP.com one, a connected JP one, or and unconnected one. Connecting and authenticating in each of those cases is an orthogonal problem IMO.

Another point I'm making here that I try to avoid domain-specific (as in scope/area, not as in internet domain name 😄) specific data structures: When setting a site title or description isn't in principle limited to the case of JPO, why artificially constrain it to that case, by e.g. throwing it into a Redux subtree that has all the things we need for JPO -- while each might be able to stand on its own feet and live in a different, more generic context (i.e. Redux subtree)?

I agree that some information that we have for WP.com or connected JP sites might just not be available for unconnected ones (heterogenity of data), and that we need to be careful not to cram data that is too heterogenous into data structures that suggest homogenity, since consumers of those structures might wrongly assume homogenity.

Maybe we need a better grasp of what kinds of actions we want to perform on those unconnected JP sites, and what kind of data we want to obtain from them to better inform this conversation. I'll have to take a closer look at the other steps and maybe compile a list of required actions.

> - when the mechanism will be working for non-connected sites, will the "connected-sites" logic be needed for this iteration at all? 

I guess it really depends a lot on what that non-connected logic will look like -- but yeah, we might end up with something that's totally connection-state-agnostic that will work in both cases, making the logic that's specific to the connected case obsolete.

> - we 'should' be displaying any site only if `ifSiteNotOnboarded` is `true`, right? Should we start thinking about having a controller on our end and including this in the logic? 

Good point. I'm hoping that this won't be too difficult -- the existing JPO plugin must already store some setting, so we'll need to either expose that (through the REST API), or move it to the Jetpack plugin itself (since we want JPO to become independent of the old plugin) and expose it from there. Plus a Calypso-side Redux reducer/selector/action that reflects/controls that setting.

tyxla commented:
Fair enough - I agree with all of your points - let's go that way, and unify the cases when we set titles/descriptions for all of the use cases. Makes sense, no need to reinvent the wheel 👍  

In that case, I think the first step here is to decouple from the `settings` bit and from the `siteId`, and set title/description separately (separate Redux actions). Actual saving/retrieving will have to be addressed separately, depending on the site type, and we can actually handle that in the data layer level.

Let's also think about where and in what format we'll store everything that we currently don't have (e.g. contact form, install WooCommerce, input business address), and how we'll store/retrieve the current flow progress (we'll need it for the summary step at the bare minimum).

ockham commented:
Rebased.

ockham commented:
Noting here that I'm putting the site fragment at the end of the route (`/jetpack/onboarding/site-title/:site`) as this is the route structure that I've generally seen suggested as more coherent (as opposed to having `:site` somewhere in between). This probably means we'll have to modify the `Wizard` component in a subsequent step.
